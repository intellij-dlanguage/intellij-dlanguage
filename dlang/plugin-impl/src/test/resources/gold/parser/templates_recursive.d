template factorial(int n : 1)
{
    enum { factorial = 1 }
}

template factorial(int n)
{
    enum { factorial = n* factorial!(n-1) }
}

void test()
{
    writefln("%s", factorial!(4)); // prints 24
}