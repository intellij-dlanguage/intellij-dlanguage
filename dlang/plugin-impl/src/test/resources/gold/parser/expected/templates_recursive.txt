D Language File
  DLanguageDeclarationImpl(DECLARATION)
    TEMPLATE_DECLARATION
      PsiElement(DlangTokenType.template)('template')
      IDENTIFIER
        PsiElement(DlangTokenType.ID)('factorial')
      DLanguageTemplateParametersImpl(TEMPLATE_PARAMETERS)
        PsiElement(DlangTokenType.()('(')
        DLanguageTemplateParameterListImpl(TEMPLATE_PARAMETER_LIST)
          TEMPLATE_PARAMETER
            DLanguageTemplateValueParameterImpl(TEMPLATE_VALUE_PARAMETER)
              DLanguageTypeImpl(TYPE)
                DLanguageType_2Impl(TYPE_2)
                  DLanguageBuiltinTypeImpl(BUILTIN_TYPE)
                    PsiElement(DlangTokenType.int)('int')
              IDENTIFIER
                PsiElement(DlangTokenType.ID)('n')
              PsiElement(DlangTokenType.:)(':')
              DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                  DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                    DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                      PsiElement(DlangTokenType.INTEGER_LITERAL)('1')
        PsiElement(DlangTokenType.))(')')
      PsiElement(DlangTokenType.{)('{')
      DLanguageDeclarationImpl(DECLARATION)
        DLanguageAnonymousEnumDeclarationImpl(ANONYMOUS_ENUM_DECLARATION)
          PsiElement(DlangTokenType.enum)('enum')
          PsiElement(DlangTokenType.{)('{')
          ENUM_MEMBER
            IDENTIFIER
              PsiElement(DlangTokenType.ID)('factorial')
            PsiElement(DlangTokenType.=)('=')
            DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
              DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                  DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                    PsiElement(DlangTokenType.INTEGER_LITERAL)('1')
          PsiElement(DlangTokenType.})('}')
      PsiElement(DlangTokenType.})('}')
  DLanguageDeclarationImpl(DECLARATION)
    TEMPLATE_DECLARATION
      PsiElement(DlangTokenType.template)('template')
      IDENTIFIER
        PsiElement(DlangTokenType.ID)('factorial')
      DLanguageTemplateParametersImpl(TEMPLATE_PARAMETERS)
        PsiElement(DlangTokenType.()('(')
        DLanguageTemplateParameterListImpl(TEMPLATE_PARAMETER_LIST)
          TEMPLATE_PARAMETER
            DLanguageTemplateValueParameterImpl(TEMPLATE_VALUE_PARAMETER)
              DLanguageTypeImpl(TYPE)
                DLanguageType_2Impl(TYPE_2)
                  DLanguageBuiltinTypeImpl(BUILTIN_TYPE)
                    PsiElement(DlangTokenType.int)('int')
              IDENTIFIER
                PsiElement(DlangTokenType.ID)('n')
        PsiElement(DlangTokenType.))(')')
      PsiElement(DlangTokenType.{)('{')
      DLanguageDeclarationImpl(DECLARATION)
        DLanguageAnonymousEnumDeclarationImpl(ANONYMOUS_ENUM_DECLARATION)
          PsiElement(DlangTokenType.enum)('enum')
          PsiElement(DlangTokenType.{)('{')
          ENUM_MEMBER
            IDENTIFIER
              PsiElement(DlangTokenType.ID)('factorial')
            PsiElement(DlangTokenType.=)('=')
            DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
              DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                DLanguageMulExpressionImpl(MUL_EXPRESSION)
                  DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                    DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                      DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                        IDENTIFIER
                          PsiElement(DlangTokenType.ID)('n')
                  PsiElement(DlangTokenType.*)('*')
                  DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                    DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                      DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                        DLanguageTemplateInstanceImpl(TEMPLATE_INSTANCE)
                          IDENTIFIER
                            PsiElement(DlangTokenType.ID)('factorial')
                          DLanguageTemplateArgumentsImpl(TEMPLATE_ARGUMENTS)
                            PsiElement(DlangTokenType.!)('!')
                            PsiElement(DlangTokenType.()('(')
                            DLanguageTemplateArgumentListImpl(TEMPLATE_ARGUMENT_LIST)
                              DLanguageTemplateArgumentImpl(TEMPLATE_ARGUMENT)
                                DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                  DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                    DLanguageAddExpressionImpl(ADD_EXPRESSION)
                                      DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                        DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                          DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                            IDENTIFIER
                                              PsiElement(DlangTokenType.ID)('n')
                                      PsiElement(DlangTokenType.-)('-')
                                      DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                        DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                          PsiElement(DlangTokenType.INTEGER_LITERAL)('1')
                            PsiElement(DlangTokenType.))(')')
          PsiElement(DlangTokenType.})('}')
      PsiElement(DlangTokenType.})('}')
  DLanguageDeclarationImpl(DECLARATION)
    FUNCTION_DECLARATION
      DLanguageTypeImpl(TYPE)
        DLanguageType_2Impl(TYPE_2)
          DLanguageBuiltinTypeImpl(BUILTIN_TYPE)
            PsiElement(DlangTokenType.void)('void')
      IDENTIFIER
        PsiElement(DlangTokenType.ID)('test')
      DLanguageParametersImpl(PARAMETERS)
        PsiElement(DlangTokenType.()('(')
        PsiElement(DlangTokenType.))(')')
      DLanguageFunctionBodyImpl(FUNCTION_BODY)
        DLanguageSpecifiedFunctionBodyImpl(SPECIFIED_FUNCTION_BODY)
          DLanguageBlockStatementImpl(BLOCK_STATEMENT)
            PsiElement(DlangTokenType.{)('{')
            DLanguageDeclarationsAndStatementsImpl(DECLARATIONS_AND_STATEMENTS)
              DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
                DLanguageStatementImpl(STATEMENT)
                  DLanguageStatementNoCaseNoDefaultImpl(STATEMENT_NO_CASE_NO_DEFAULT)
                    DLanguageExpressionStatementImpl(EXPRESSION_STATEMENT)
                      DLanguageExpressionImpl(EXPRESSION)
                        DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                          DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                            DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                              DLanguageFunctionCallExpressionImpl(FUNCTION_CALL_EXPRESSION)
                                DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                  DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                    DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                      IDENTIFIER
                                        PsiElement(DlangTokenType.ID)('writefln')
                                DLanguageArgumentsImpl(ARGUMENTS)
                                  PsiElement(DlangTokenType.()('(')
                                  DLanguageArgumentListImpl(ARGUMENT_LIST)
                                    DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                      DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                        DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                          DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                            PsiElement(DlangTokenType.DOUBLE_QUOTED_STRING)('"%s"')
                                    PsiElement(DlangTokenType.,)(',')
                                    DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                      DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                        DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                          DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                            DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                              DLanguageTemplateInstanceImpl(TEMPLATE_INSTANCE)
                                                IDENTIFIER
                                                  PsiElement(DlangTokenType.ID)('factorial')
                                                DLanguageTemplateArgumentsImpl(TEMPLATE_ARGUMENTS)
                                                  PsiElement(DlangTokenType.!)('!')
                                                  PsiElement(DlangTokenType.()('(')
                                                  DLanguageTemplateArgumentListImpl(TEMPLATE_ARGUMENT_LIST)
                                                    DLanguageTemplateArgumentImpl(TEMPLATE_ARGUMENT)
                                                      DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                                        DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                                          DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                                            DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                                              PsiElement(DlangTokenType.INTEGER_LITERAL)('4')
                                                  PsiElement(DlangTokenType.))(')')
                                  PsiElement(DlangTokenType.))(')')
                      PsiElement(DlangTokenType.;)(';')
            PsiComment(DlangTokenType.LINE_COMMENT)('// prints 24')
            PsiElement(DlangTokenType.})('}')